Cabal and Cabal-syntax API checking

geekosaur opened the pull request:
Not checking cabal-install or cabal-install-solver currently because they don't have public APIs. This may be incorrect for cabal-install-solver, so that may be added later.

**Template B: This PR does not modify behaviour or interface**

*E.g. the PR only touches documentation or tests, does refactorings, etc.*

Include the following checklist in your PR:

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions).
* [x] Is this a PR that fixes CI? If so, it will need to be backported to older cabal release branches (ask maintainers for directions).

There should be no need for manual QA, because it's already been tested by multiple rebases on `master` which changed API (for example, https://github.com/haskell/cabal/pull/10270).

ulysses4ever commented:
Wow, this is awesome! :star_struck:

geekosaur commented:
Do you happen to know whether cabal-install-solver needs to be included here, since it's a library? My assumption is that it's internal and as such has no _public_ API to be checked.

geekosaur commented:
Also, this is very much WIP as I sort out what's needed (and deal with weirdshit like the fact that I somehow got a byte order mark the first time, and subsequent runs don't like it much).

ulysses4ever commented:
Cabal and Cabal-syntax are mission critical. The solver package is not (e.g. reverse dependencies count from Hackage: 259 for Cabal vs 1 for solver). So it's fine to start with those.

geekosaur commented:
Just made the current API dumps into artifacts, since not everyone will have an Ubuntu box to run `print-api` on and sufficiently large changes would be a PITA working from diff output.

geekosaur committed:
first cut at API checking

It turns out there is a consumer of cabal-install-solver, so I have
added it to API generation and checking.

Kleidukos commented:
@geekosaur why not use GHC 9.10.1? :)

geekosaur commented:
Originally I was thinking 9.4.8 for consistency with GHC_FOR_RELEASE, then discovered that wasn't supported so I bumped it. I could do it again with 9.10.1 if that's preferred.

Kleidukos commented:
Yes, let's use the highest supported GHC now and bump it once it falls out of the support window

mpickering reviewed (dismissed):
Please can we provide a means to easily update these files with a single command locally? 

Ideally it should be possible to run all cabal tests locally with a single command and update the output within the same framework. This is already an issue with the structured hash test, which is difficult to update and dependent on GHC version, and in practice, just more annoying busywork than informative for contributors. 

Otherwise, I don't think this job should be required by contributors as it introduces a significant amount of friction. It could be run as part of CI to track how the API has changed over time rather than requiring a golden file to update or perhaps this test could just be run on release branches (where the API is not expected to change).

sheaf commented:
I think having such an API test is going to be very valuable, as it allows us to confidently assess the API impact of PRs under review. This will give us much more confidence and allow us to write more accurate changelog, and avoid accidental changes in minor Cabal version etc. I'm grateful that the Cabal project is taking such a step.

However, I share Matthew's concern about the workflow for accepting the test. At the moment one has to provision a docker container to run with the appropriate environment for the test. Either that, or one ends up with a situation like the structured hash test where one has to wait for CI to run, let it fail because of the test, then change the stdout locally, then push again. If one needs one more CI run than one did before then I think it is too onerous; I would much rather that the structured hash test and this test be separate tests which don't cause the whole CI pipeline to fail if they fail, so that they can be updated in parallel with working on the PR without cancelling the CI which might otherwise have valuable information.

I also thing the surface area we are testing is a bit too large (I don't think the Cabal project has any guarantees about the entire set of exposed modules; that seems too much). As a result, I'm worried that we will get unreadable diffs as a result whenever there is a significant refactor. But I suppose we can wait and see how things pan out in that regard.

Kleidukos commented:
> I think having such an API test is going to be very valuable, as it allows us to confidently assess the API impact of PRs under review. This will give us much more confidence and allow us to write more accurate changelogs, and avoid accidental changes in minor Cabal version etc. I'm grateful that the Cabal project is taking such a step.

:+1: 

> However, I share Matthew's concern about the workflow for accepting the test

We all do, I don't think anyone has objected publicly or privately.

> I also thing the surface area we are testing is a bit too large (I don't think the Cabal project has any guarantees about the entire set of exposed modules; that seems too much). As a result, I'm worried that we will get unreadable diffs as a result whenever there is a significant refactor. But I suppose we can wait and see how things pan out in that regard.

Yep, configuring the tool with an "ignore" list is in the plans. Cabal is just an early adopter and the needs of adopters will help determining how `print-api` & `diff-package-api` can provide the most value. Example: @geekosaur had to bypass the pre-made Workflow in order to test multiple packages at once.

geekosaur commented:
> However, I share Matthew's concern about the workflow for accepting the test. At the moment one has to provision a docker container to run with the appropriate environment for the test.

Actually no: as a first cut, the new API files are artifacts which can be downloaded and copied over the existing ones, specifically for this case. If there's some friendly way to make it completely automatic, I'm all ears, but I did specifically address that "contributors need to replicate the CI environment" issue. (And will shortly be using it myself, since I'll be redoing the golden API files for 9.10.1.)

geekosaur commented:
> Otherwise, I don't think this job should be required by contributors as it introduces a significant amount of friction. It could be run as part of CI to track how the API has changed over time rather than requiring a golden file to update or perhaps this test could just be run on release branches (where the API is not expected to change).

As semi-official release manager for the 3.12 branch, I would like this to run on every incoming PR. I do need to tune that though, as I don't want API changes to prevent merges; I want them to inform me whether a given PR is a backport candidate or not. As I'm still learning the details of workflows, I should probably put this back in draft even though I'm soliciting reviews.

geekosaur commented:
Uh, let me correct that: the test should always pass, this is accomplished by updating the golden tests by copying the new API files over the old ones as described above. Then the presence of such updates tells me as release manager that the PR shouldn't be backported.

I think my ideal for both this and the structured hashes is that they'd be checkboxes in the PR that the tests could inspect in order to determine whether to report a mismatch or do an update. But
1. AFAICT there's no way to inspect PR checkboxes from workflows;
2. the multiple templates would make this difficult anyway.

geekosaur commented:
Okay, I've added documentation about the check API job and how to update the API in a PR. The actual API check also outputs a note about where to find the new API artifacts.

geekosaur commented:
Also, I switched out `diff` for `cmp`, because with `Cabal`'s 1MB+ API footprint nobody really cares about the details, just that it changed. If someone really wants it, they can download the new API files and run `diff` locally.

geekosaur commented:
> Ideally it should be possible to run all cabal tests locally with a single command and update the output within the same framework.

The problem with this is that GHC and `base` internals (in particular, basic types and typeclasses) are part of the API, which means the API depends on the GHC version and to some extent the platform `print-api` is run on. As such, contributors (unless they happen to be on Ubuntu 22.04 on x86_64) would need a Docker container, `virtualbox`, or similar to reproduce the exact API. (It can be argued that that's not strictly part of **`Cabal`'s** API, but then we run into `print-api` not having a way to suppress such internals as yet.)

This is why I only check the API on a single platform and GHC version, and have tried to make it as easy as possible to update the golden API record when needed.

ulysses4ever commented on .github/workflows/check-api.yml:
```diff
@@ -0,0 +1,93 @@
+      - name: Check Cabal-syntax and Cabal APIs
+        run: |
+          rc=0
+          if cmp -s Cabal-syntax/Cabal-syntax-${{ matrix.ghc }}-${{ matrix.sys.os }}.api Cabal-syntax-${{ matrix.ghc }}-${{ matrix.sys.os }}.api; then
```
I feel like the decision to opt for a simpler check with `cmp` rather than going full`diff` is not the right choice for CI in particular. You say: 

> switched out diff for cmp, because with Cabal's 1MB+ API footprint nobody really cares about the details, just that it changed. 

So is the only reason for not doing diff is that no one cares?  I think I may care in some reasons. Is there a more objective reason? Does it save time or something? One could run `cmp` first and only if that fails run `diff`...

geekosaur replied:
Maybe I'm overreacting to the first run, where it was `diff`ing against an empty file and so dumped the entire API. It was pretty much impenetrable, unorganized, and managed to break GitHub's log viewer so I ended up downloading the build log to read it.

Some of this is just that `print-api` output isn't really designed for humans. Much of it is that we just have too much API, in particular in `Cabal`. (Diffs work for `text-display`, but `print-api`'s entire output is about 12 lines there. `Cabal`'s is 1000 times larger, even if you remove its re-export of `Cabal-syntax`.)

In re "run `cmp` first and only if that fails run `diff`", both produce no output if they succeed so there would be no difference. Possibly we could run `diff` to a file and output it if it's small enough, but then we will bikeshed over what counts as "small enough".

geekosaur replied:
I've switched to `diff` (which shortly will be `diff -c` because plain `diff` output is a little too obtuse) and show the changes if there's fewer than 50 lines. We can tune this to some extent, likely limited by what GitHub's log viewer can cope with.

ulysses4ever commented on .github/workflows/check-api.yml:
```diff
@@ -0,0 +1,93 @@
+          ./print-api --package-name Cabal-syntax > Cabal-syntax-${{ matrix.ghc }}-${{ matrix.sys.os }}.api
+          ./print-api --package-name Cabal > Cabal-${{ matrix.ghc }}-${{ matrix.sys.os }}.api
+
+      # for convenience, since large changes would be a pain to reconstruct from diffs and
```
It's really hard for me to follow this whole comment. What's it trying to tell me? Any chance it could state its purpose more clearly?

ulysses4ever commented on CONTRIBUTING.md:
```diff
@@ -193,6 +193,20 @@ We use automated whitespace convention checking. Violations can be fixed by
+-----------
+
+The `Check API` job tests the `Cabal` and `Cabal-syntax` packages for API changes.
+It's still something of a work in progress, but even in its current state it is
```
I feel like "work in progress" may suggest someone that it can be ignored. Since I don't see what this phrase achieves, I'd not put it here. Generally, if the idea won't work as smooth as we hope, we can always nuke it even if it's not a WIP...

geekosaur replied:
My intent was to suggest that this section is subject to change and contributors should check back to see if it has — but I expect such changes if any to happen relatively quickly.

ulysses4ever commented on CONTRIBUTING.md:
```diff
@@ -193,6 +193,20 @@ We use automated whitespace convention checking. Violations can be fixed by
 running [fix-whitespace](https://hackage.haskell.org/package/fix-whitespace). If
 you push a fix of a whitespace violation, please do so in a _separate commit_.
 
+API Changes
```
```suggestion
API Changes and Check API job
```

ulysses4ever reviewed (approved):
I understand reservations about more cumbersome contributing process. But we don't fully understand in advance how it will pan out. It's something of an innovation, and I'm eager to see how it works in practice. If it ends up endless pain, we should reconsider. 

I agree that having no way to update it locally is unfortunate. I think the documentation should provide some basic solution. Even if it depends on Linux, version of GHC, and even some particular Linux distro (why??). Those who really hate to round-trip to CI will find a way to adapt a basic instruction.

geekosaur commented:
@ulysses4ever, I'm not currently finding your question about API vs. platform.
1. Different Linux distributions on the same architecture shouldn't have different APIs, but might require using a different `print-api` release package because of `glibc` incompatibilities (the `print-api` release only includes one shared version for Linux, and the example usage on `text-display` strongly suggests it's for Ubuntu 22.04).
2. MacOS and Windows on the same architecture probably differ, because they have different ABIs and this can potentially show up in some of ghc's/`ghc-prim`'s/`ghc-base`'s types. Ideally `print-api` would exclude (or use placeholders for) these types, but I suspect that's firmly in `ghc-api` territory and may make `print-api` as difficult to support across GHC versions as HLS is (that is, sometimes breaking across minor versions that aren't supposed to affect normal programmer-visible API/ABI).
3. Different architectures almost certainly will differ, for the same reason as (2). `CInt` on AArch64 vs. x86_64 is an example here: on x86_64 the decision was made to keep C's `int` 32 bit for backward compatibility, but IIRC AArch64 went with 64 bits, so any API using `CInt` (not that I expect `Cabal` to have such, hopefully!) would differ because the type alias will expand to `Word32` vs. `Word64`.

geekosaur commented:
Some things to consider for the future:
* I have trouble believing `Cabal`'s API is that large. How much of it should move to an internal library? (Having `print-api` ignore things defined in `Cabal` would be a workaround.)
  - But beware of things that people depend on because they're considered part of the API currently (see for example #9856)
* Dependencies can have `Internal` modules, which are excluded from the public API but still visible. `print-api` would need an option to ignore these, or possibly an annotation similar to the ones in Haddock (`hide`, `not-home`) — which might even be reusable here.

mpickering commented:
I am against merging anything which requires downloading files from CI in order to make tests pass. This workflow is much too cumbersome for developers. The structured hash check is already bad enough when it comes to PRs and having to do multiple rounds.

If the goal is to stop API changes happening on stable branches, then it seems that running this job only on release branches is the way to go. At which point the test should never/rarely fail and the burden is placed on the release manager (who desires this check) to update the files rather than the developer whose goal is to fix a bug or write a feature.

geekosaur commented:
> I am against merging anything which requires downloading files from CI in order to make tests pass

Are you prepared to confirm that ghc doesn't force this on us by producing different types on different platforms?

Bodigrim commented:
> 1. Different Linux distributions on the same architecture shouldn't have different APIs, but might require using a different `print-api` release package because of `glibc` or other shared library incompatibilities

It seems to me that `print-api` is small enough to be built from sources every time instead of relying on pre-built binaries.

> 3\. Different architectures almost certainly will differ, for the same reason as (2). `CInt` on AArch64 vs. x86_64 is an example here

I don't quite follow this example: `CInt` is `CInt`, it's not a type synonym. Does `print-api` unwrap it for some reason?

Kleidukos commented:
@Bodigrim I can give an answer for point N°2: CInt ought not to change for consumers of the type. Only its place of definition will produce different results on different platforms.

See the reason written by Ben Gamari here: https://github.com/Kleidukos/print-api/blob/main/src/PrintApi/IgnoredDeclarations.hs#L37

See an example of "redacted" definition: https://gitlab.haskell.org/ghc/ghc/-/blob/master/testsuite/tests/interface-stability/base-exports.stdout-javascript-unknown-ghcjs#L2141-2142

Bodigrim commented:
> CInt ought not to change for consumers of the type. Only its place of definition will produce different results on different platforms.

I still don't quite get it: as much as I can tell `CInt` is defined [in the same place](https://hackage.haskell.org/package/ghc-internal-9.1001.0/docs/src/GHC.Internal.Foreign.C.Types.html#CInt), independently of a platform. The RHS of the definition is platform-dependent indeed, but unless you run `print-api` on `base` itself, this should not be a concern, if I understand it correctly.

Kleidukos commented:
@Bodigrim That is absolutely what I meant, indeed. :)

geekosaur commented:
> It seems to me that `print-api` is small enough to be built from sources every time instead of relying on pre-built binaries.

If the objective is to minimize extra work needed by contributors, I'm not sure that's the way to go. Ideally they'd be able to `cabal install print-api`.

geekosaur commented:
The current version of this has `Makefile` rules to generate the API records locally. (You need to install `print-api` yourself.) If the GHC ecosystem behaves itself then this should be acceptable.

mpickering commented:
Perhaps `print-api` could be packaged as an external command, so that `cabal print-api` would deal with building and installing the executable, building the project appropiate and so on.

(For example, `cabal-doctest`, https://github.com/sol/doctest/blob/main/src/Cabal.hs)

mpickering commented:
Thanks for the `Makefile` target, that's an improvement. I don't know if the output will be invariant across platforms. I had a brief look but I think that the scope of the checking needs to be dialed down a bit.. for example `Distribution.Compat.Binary` is clearly supposed to be an "internal" module and just reexports some definitions for convenience within the library. Another one with a lot of reexports is `Distribution.Compat.Prelude`.

Kleidukos commented:
@mpickering This is exactly what the next release of the `print-api` will address, by enabling end-users to maintain an "ignore list" of modules. Another future version will even be enabled to work with Haddock module attributes.
This PR will use the next version before getting merged, although discussion about ergonomics is important and informs the development of `print-api`.

geekosaur commented:
I also have started a discussion somewhere (maybe above) about reviewing Cabal APIs that belong in private libraries instead of the public one, but I suspect that's a long term goal. (Oh, it's [the first item here](https://github.com/haskell/cabal/pull/10259#issuecomment-2295348038).)

geekosaur commented:
Also, that upcoming ignore list is why I moved the API generation into the `Makefile` instead of showing it directly in `CONTRIBUTING.md` as in an earlier revision of this patch; it'll definitely be asking too much to have contributors track possibly long ignore lists.

geekosaur commented:
> Another one with a lot of reexports is `Distribution.Compat.Prelude`.

I have a nasty suspicion that other people are using these modules to piggyback on our backward compatibility mechanisms. After all, it's a bootlib.

Bodigrim commented:
> I have a nasty suspicion that other people are using these modules to piggyback on our backward compatibility mechanisms. After all, it's a bootlib.

I don't think there is any significant usage of `Distribution.Compat.Prelude` outside of Cabal packages: https://hackage-search.serokell.io/?q=Distribution.Compat.Prelude
People usually pick `base-compat` for this purpose.

OTOH if the goal is to determine whether a PR changes API at all or not, it does count, unfortunately.

Kleidukos commented:
Let's discuss the scope of the supported lib:Cabal API in another ticket, I feel like there is more to plan.

geekosaur commented:
I did say it was long term…

geekosaur commented:
#10276 will fix this.

geekosaur commented:
It appears that I'll need to add `cabal-install-solver` after all: there's a Nix-related tool using it, according to Hackage.

geekosaur commented:
@Kleidukos, is there a way to tell `print-api` to use a different ghc? It turns out I can't run it on my system without switching my default ghc to 9.10.1 (it's normally 9.6.6). Everything else I can pass `-w ghc-9.10.1` to and it works as expected.

geekosaur commented:
Turns out I lost the last batch of changes (forgot to unstash them after rebasing other PRs) so I ended up redoing them, then discovered that the `Makefile` rules wouldn't work as written because `make` doesn't know the dependencies, `cabal` does.

There is still one issue with local API generation, which is that `print-api` assumes your default `ghc` is the one it's linked against.

Kleidukos commented:
> @Kleidukos, is there a way to tell `print-api` to use a different ghc? It turns out I can't run it on my system without switching my default ghc to 9.10.1 (it's normally 9.6.6). Everything else I can pass `-w ghc-9.10.1` to and it works as expected.

Yes, we are still figuring out the ergonomics with @mmhat. He found out that using `ghcup run --ghc 9.10.1 -- <your command>` works. Does that work for you as a temporary solution?

geekosaur commented:
@Kleidukos, @ffaf1, not sure if this is something in `print-api` or something we missed in #10309, but I just got the following:
```
print-api: <no location info>: error: [GHC-87110]
    Could not load module `Distribution.Backpack'.
    It is a member of the hidden package `Cabal-syntax-3.13.0.0'.
HasCallStack backtrace:
  finally, called at compiler/GHC/Utils/Panic.hs:303:7 in ghc-9.10.1-69c3:GHC.Utils.Panic
```
Since I just rebased on `master`, this should want 3.15.0.0, no?

geekosaur commented:
(despite the labels, this is parked until post-ghc-9.12.1)

If/when this _does_ go in, "Check API Post Job" will need to be added to the required jobs.

ulysses4ever commented on Makefile:
```diff
@@ -174,6 +174,68 @@ cabal-install-test-accept:
 	rm -rf .ghc.environment.*
 	cd cabal-testsuite && `cabal list-bin cabal-tests` --with-cabal=`cabal list-bin cabal` --hide-successes -j3 --accept ${TEST}
 
+# API generation
```
For this much code we should start using gigantic headers to facilitate navigation. E.g.
```
##########################################################################
#                                                                        #
#                              API generation                            #
#                                                                        #
##########################################################################
```
or something...

This applies to other sections as well of course, and is long overdue.

ulysses4ever replied:
fair

philderbeast commented:
@geekosaur do you want to split this into two PRs? It would be useful to run the make targets locally if we suspect an API change.

geekosaur commented:
@philderbeast, I'm not sure what a split would be worth. The only split that makes any sense to me would be to make the API files a separate commit in this PR, facilitating cherry-picking the rest into a test branch.

philderbeast commented:
@geekosaur, I'm trying this out. I was able to install `print-api` with `make print-api-install`:

```makefile
PRINT_API_VERSION ?= eedf83e6f828217ba7946ca8bfaf4ab4062c2363
PRINT_API_URL := https://github.com/Kleidukos/print-api/archive/${PRINT_API_VERSION}.tar.gz

print-api/print-api.cabal:
	rm -rf print-api
	curl -sSL ${PRINT_API_URL} | tar -xz
	mv print-api-* print-api

print-api-install: print-api/print-api.cabal
	cabal install print-api:exe:print-api --project-dir=print-api --overwrite-policy=always
```

philderbeast commented:
@geekosaur I was thinking of a split into two, one with the makefile changes (so we could use it locally) and the other with CI changes.

philderbeast commented:
When I run it I'm seeing an error:

```
$ make generate-api-cabal-syntax
...
ghcup run --ghc 9.10.1 -- print-api --package-name Cabal-syntax | \
   sed 's/\([( ]\)[Cc]abal-[-0-9.][-0-9]*:/\1/g' >Cabal-syntax-9.10.1.api
Detected GHC version: 9.10.1
print-api: print-api-9.10.1: can't find a package database at /.../dist-newstyle/packagedb/ghc-9.10.1
HasCallStack backtrace:
  finally, called at compiler/GHC/Utils/Panic.hs:303:7 in ghc-9.10.1-69c3:GHC.Utils.Panic
```

geekosaur commented:
Huh. I guess I'll have to add that to the rules. (I seem to have hit the same problem when trying to sanity check 3.12 branch against its base.)

geekosaur commented:
Uhhhhh… @Kleidukos, why is print-api trying to build cabal-testsuite? (This is on 3.12.1.0, and it looks like something got bumped inappropriately.)
```
ghcup run --ghc 9.10.1 -- print-api --package-name cabal-install-solver | sed 's/\([( ]\)[Cc]abal-[-0-9.][-0-9]*:/\1/g' >cabal-install-solver-9.10.1.api
print-api: Error: [Cabal-7107]
Could not resolve dependencies:
[__0] trying: cabal-testsuite-3 (user goal)
[__1] next goal: cabal-testsuite:setup.Cabal (dependency of cabal-testsuite)
[__1] rejecting: cabal-testsuite:setup.Cabal-3.14.0.0 (conflict: cabal-testsuite => cabal-testsuite:setup.Cabal>=3.10 && <3.11)
[__1] skipping: cabal-testsuite:setup.Cabal; 3.12.1.0, 3.12.0.0/installed-be90, 3.12.0.0 (has the same characteristics that caused the previous version to fail: excluded by constraint '>=3.10 && <3.11' from 'cabal-testsuite')
[__1] rejecting: cabal-testsuite:setup.Cabal; 3.10.3.0, 3.10.2.1, 3.10.2.0, 3.10.1.0, 3.8.1.0, 3.6.3.0, 3.6.2.0, 3.6.1.0, 3.6.0.0, 3.4.1.0, 3.4.0.0, 3.2.1.0, 3.2.0.0, 3.0.2.0, 3.0.1.0, 3.0.0.0, 2.4.1.0, 2.4.0.1, 2.4.0.0, 2.2.0.1, 2.2.0.0, 2.0.1.1, 2.0.1.0, 2.0.0.2, 1.24.2.0, 1.24.0.0, 1.22.8.0, 1.22.7.0, 1.22.6.0, 1.22.5.0, 1.22.4.0, 1.22.3.0, 1.22.2.0, 1.22.1.1, 1.22.1.0, 1.22.0.0, 1.20.0.4, 1.20.0.3, 1.20.0.2, 1.20.0.1, 1.20.0.0, 1.18.1.7, 1.18.1.6, 1.18.1.5, 1.18.1.4, 1.18.1.3, 1.18.1.2, 1.18.1.1, 1.18.1, 1.18.0, 1.16.0.3, 1.16.0.2, 1.16.0.1, 1.16.0, 1.14.0, 1.12.0, 1.10.2.0, 1.10.1.0, 1.10.0.0, 1.8.0.6, 1.8.0.4, 1.8.0.2, 1.6.0.3, 1.6.0.2, 1.6.0.1, 1.4.0.2, 1.4.0.1, 1.4.0.0, 1.2.4.0, 1.2.3.0, 1.2.2.0, 1.2.1, 1.1.6, 1.24.1.0 (constraint from minimum version of Cabal used by Setup.hs requires >=3.12)
[__1] fail (backjumping, conflict set: cabal-testsuite, cabal-testsuite:setup.Cabal)
After searching the rest of the dependency tree exhaustively, these were the goals I've had most trouble fulfilling: cabal-testsuite:setup.Cabal, cabal-testsuite


CallStack (from HasCallStack):
  error, called at src/PrintApi/CLI/Types.hs:34:7 in print-api-0.1.0.1-inplace:PrintApi.CLI.Types
HasCallStack backtrace:
  collectBacktraces, called at libraries/ghc-internal/src/GHC/Internal/Exception.hs:92:13 in ghc-internal:GHC.Internal.Exception
  toExceptionWithBacktrace, called at libraries/ghc-internal/src/GHC/Internal/Exception.hs:128:3 in ghc-internal:GHC.Internal.Exception
```

geekosaur commented:
Welp. With this fix it appears that a few API changes leaked into 3.12 after all. That said, they're slight and I think they might be safe because there's backward compatibility code.

- `Cabal-syntax`: new pattern synonym, which provides backward compatibility to changes to `SCC`, plus a `Foldable1` instance for it
- `Cabal`: reexports of the preceding
- `cabal-install-solver`: reexporting the above `Foldable1` instance

ulysses4ever commented:
> new pattern synonym, which provides backward compatibility to changes to SCC, plus a Foldable1 instance for it

Note PVP:

> A package version number SHOULD have the form A.B.C ... if only new bindings, types, classes, non-orphan instances or modules (but see below) were added to the interface, then A.B MAY remain the same but the new C MUST be greater than the old C

https://github.com/haskell/pvp/blob/master/v1.0/pvp-specification.md

geekosaur commented:
Yes, but this would be 3.14.2.0 (C increase). The patsyn shouldn't count since it is maintaining the 3.14.1.0 API; the addition of the replacement constructor to `SCC` _with_ the backward compatibility patsyn should then also be a minor version bump (C), I hope. If not, I'll have to track down what commit added the change and revert it.

geekosaur commented:
Oh, this is in a dependency (`Data.Graph`, used by `Cabal-syntax:Distribution.Compat.Graph). Does that mean tracking down a version that doesn't have these, or can we just eat the change?

Kleidukos commented:
> (I had assumed that `SCC` meant GHC, whoops.)

Ah, Strongly Connected Components are **not** Stack Cost Centres :P

Kleidukos commented:
> Does that mean tracking down a version that doesn't have these, or can we just eat the change?

The deal with compat modules somewhat eludes me, we need to either ask people who set them in place, or actually ask people who use them (hopefully hackage-search can help us), and ask the end-users for their expectations.

geekosaur commented:
Yeh, I was wondering why we'd be looking at that level of detail when we only care about enabling profiling. 😀

geekosaur commented:
I'm not sure we even care about the change as such; I bet if we search the codebase it'll still be using the compatibility patsyn. ezyang hasn't made any commits since adding Backpack support, has he? Certainly not since 3.12.

geekosaur commented:
Right, nothing uses the new constructor, or even the old one.

Mikolaj commented:
Let me unmark this in order not to block the release, but if we tie any loose ends before the nearest minor 3.14 release, let's re-add the backport label.

ulysses4ever commented:
I just learned from a recent Cafe thread that the streamly folks created their own tool for API checking based off of haddock/hoogle output (which sounds very natural): https://github.com/composewell/packdiff maybe we should dig into it?..
